### PR TITLE
rough carousel

### PR DESCRIPTION
**Pull request by danmarshall**
the styling leaves a lot to be desired, but some basics are sketched in

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
   },
   "devDependencies": {
     "http-server": "^0.9.0",
+    "source-loader": "^0.2.0",
```
Comment (billba) [inline package.json]: Do we need this anymore? I'm running fine without this.

File: src/Carousel.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        };
+
+    }
+
+    private componentDidMount() {
+        var li = this.state.ul.firstChild as HTMLLIElement;
+        this.state.itemWidth = li.offsetWidth;
+    }
+
+    private scrollBy(increment: number) {
```
Comment (billba) [inline src/Carousel.tsx]: I'm not necessarily disagreeing with this strategy, but I'd love you to explain to me why this is all going on here and not just adjusting state for successive renders.

File: src/Carousel.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        this.state = {
+            scrollIndex: 0,
+            itemWidth: null
+        };
+
+    }
+
+    private componentDidMount() {
+        var li = this.state.ul.firstChild as HTMLLIElement;
```
Comment (billba) [inline src/Carousel.tsx]: const (or just inline this as an expression in the next line)

File: src/Carousel.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    private scrollBy(increment: number) {
+        //this.state.scrollIndex -= increment;
+        //this.state.scrollDiv.scrollLeft += increment * this.state.itemWidth;
+        //this.setState(this.state);
+
+        //TODO: cancel existing animation when clicking fast
+        //TODO: animate to boundaries
+
+        var unit = increment * this.state.itemWidth;
```
Comment (billba) [inline src/Carousel.tsx]: const

File: src/Carousel.tsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+        var div = document.createElement('div');
```
Comment (billba) [inline src/Carousel.tsx]: const

File: src/Carousel.tsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        div.className = 'wc-animate-scroll';
+        div.style.left = this.state.scrollDiv.scrollLeft + 'px';
+        document.body.appendChild(div);
+
+        //capture ComputedStyle every millisecond
+        var captureTimer = setInterval(() => {
```
Comment (billba) [inline src/Carousel.tsx]: const

File: src/Carousel.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        var finishTimer = setTimeout(() => {
+            clearInterval(captureTimer);
+            document.body.removeChild(div);
+        }, 1000 * parseFloat(getComputedStyle(div).transitionDuration) || 5000);
+
+    }
+
+    render() {
+
+        let items = this.props.attachments.map((attachment, i) => {
```
Comment (billba) [inline src/Carousel.tsx]: const

File: src/Carousel.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            return <li><Attachment attachment={attachment} actions={this.props.actions}/></li>;
```
Comment (billba) [inline src/Carousel.tsx]: omit the brackets, return, and semicolon

File: src/Carousel.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        }) as React.ReactElement<any>[];
```
Comment (billba) [inline src/Carousel.tsx]: do you need this type coercion at the end?

File: src/History.tsx [hunk 10]
```diff
@@ -4,7 +4,7 @@ import { Timestamp } from './Timestamp.tsx';
 import { MessageGroup, HistoryActions } from './App.tsx';
 import { HistoryMessage } from './HistoryMessage.tsx';
 
-interface Props {
+export interface Props {
```
Comment (billba) [inline src/History.tsx]: Does this actually help anything? At some point I tried this and it didn't seem to make a difference.

Ditto with the other props you exported below.

File: tsconfig.json [hunk 11]
```diff
@@ -1,22 +1,23 @@
...
-        "src/directLine.ts",
-        "src/mockLine.ts",
-        "src/HeroCard.tsx",
-        "src/Timestamp.tsx",
-        "src/HistoryMessage.tsx",
-        "src/Console.tsx",
-        "src/History.tsx",
-        "src/App.tsx"
-    ]
+  "compilerOptions": {
```
Comment (billba) [inline tsconfig.json]: what happened here? Did the indentation go away?

File: src/Carousel.tsx (hunk 2 above)
Comment (danmarshall) [inline src/Carousel.tsx]: Generally I'm trying to leverage the native scrolling and not reinvent it. I'm using the scrollLeft property which is driven by resize. This is still being worked out.

**Pull request merged**